[develop] Updated option to use PLATFORM=container instead of singularity (#1373)
ufs-community/ufs-srweather-app: doc/UsersGuide/BuildingRunningTesting/ContainerQuickstart.rst

large dependency software stacks on every machine, reduces setup time, and supports
consistent workflows across different systems and cloud platforms.

This section distinguishes the following container workflows:

* **Intel-based pre-built SRW runtime container workflow:** 

>   This workflow uses a container that includes a pre-built SRW App v3.0

Reviewer: There is no page break in the RtD documentation.  It would be best to change This to this.

```suggestion
* **Intel-based pre-built SRW runtime container workflow:** 
  this workflow uses a container that includes a pre-built SRW App v3.0
```

  executable and its runtime environment. The pre-built application is staged
  from the container and then used to run the community test case.

* **Containerized software-stack workflow for building and running the SRW App:** 

>   This workflow uses a container that provides the software stack needed to

Reviewer: Similar to above, there is no page break in the RtD documentation. It would be best to change This to this.

```suggestion
* **Containerized software-stack workflow for building and running the SRW App:** 
  this workflow uses a container that provides the software stack needed to
```

  build and run the UFS SRW App from source. The first step depends on the
  container option being used. Users may use a staged GNU-based or Intel-based
  software-stack container available on Tier 1 NOAA RDHPC platforms, or they may
  build a GNU-based or Intel-capable Singularity/Apptainer image from Docker

>   Hub on another supported system. After the container is available, the

Reviewer: Since the container is available on Tier-1 platforms, and we only support Tier-1 platforms, there would be no need to build either a GNU or Intel image on another supported system.

```suggestion
  Hub on non-supported systems. After the container is available, the
```

   For detailed build instructions without containers, see :numref:`Section %s <BuildSRW>`.

.. _PrerequsitesC:

-------------------
Prerequisites 
-------------------

The following prerequisites apply to **all** container workflows.

.. _ContainerSoftrareInstallC:

Singularity/Apptainer Installation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Users must have **Singularity** or **Apptainer** installed on their compute platform. 

On many HPC systems, Singularity/Apptainer may be available as a loadable
module:

.. code-block:: console

   module load singularity
   # or
   module load apptainer

When not available system-wide, Apptainer could be installed on Linux-based
system following `Apptainer Installation Guide
<https://apptainer.org/docs/admin/latest/installation.html>`__. 
This will include the installation of all dependencies. 

Further information on Singularity/Apptainer is available at:

- Singularity/Apptainer container solution for HPC systems:
  `https://en.wikipedia.org/wiki/Apptainer#History
  <https://en.wikipedia.org/wiki/Apptainer#History>`_

- SingularityCE:
  `https://sylabs.io/singularity/ <https://sylabs.io/singularity/>`_
- SingularityCE Documentation:

>   `https://docs.sylabs.io/ <https://docs.sylabs.io/>`_

Reviewer: The `doc_tests` GHA is failing.  This URL is being redirected to https://sylabs.io/docs/.

```suggestion
  `https://sylabs.io/docs/ < https://sylabs.io/docs/ >`_
```

   and local configuration.

Some platforms provide Singularity/Apptainer by default. Others require a module
load before building or running the container.

.. list-table:: Container software used on NOAA RDHPC Tier 1 platforms
   :widths: 25 25 30
   :header-rows: 1

   * - Machine
     - Container command
     - Module to load
   * - Ursa
     - ``apptainer``
     - none required
   * - Gaea
     - ``apptainer``
     - none required
   * - Hercules/Orion
     - ``singularity``
     - ``module load singularity``
   * - Derecho
     - ``apptainer``
     - ``module load apptainer``
   * - NOAA Cloud AWS/Azure
     - ``singularity``
     - none required
  
.. _CompilerMPIReqC:

Compiler and MPI Requirements
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Although containers may provide a complete SRW software stack or the software
libraries needed to build the SRW App, runtime execution still depends on
compatible MPI support on the host system. In the Tier 1 platform
examples,  Slurm launches MPI tasks the host system. The
host-side MPI startup then communicates with the binary-compatible MPI library
inside the container: Intel MPI for Intel-based containers, or OpenMPI built

> with PMI2 support for GNU-based containers . On unsupported systems,

Reviewer: ```suggestion
with PMI2 support for GNU-based containers. On unsupported systems,
```

Apptainer is fully compatible with Singularity, and commands shown here using ``singularity`` may be 
replaced with ``apptainer`` as appropriate.

On many HPC systems, Singularity/Apptainer may be available as a loadable
module:

.. code-block:: console

   module load singularity
   # or
   module load apptainer

When not available system-wide, Apptainer could be installed on Linux-based
system following `Apptainer Installation Guide
<https://apptainer.org/docs/admin/1.2/installation.html>`__. 
This will include the installation of all dependencies. 

Compiler and MPI Requirements
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Although containers may provide a complete SRW software stack or software libraries
to build the SRW App, runtime execution still depends on compilers and/or
binary-compatible MPI implementation on the host system.

* The **Intel-based container** requires Intel compilers and Intel MPI (or the
  Intel oneAPI toolkit).

> * The **GNU-based container** may need GNU compilers (GCC 12+ recommended), an

Reviewer: Should be a binary, rather than an binary.

```suggestion
* The **GNU-based container** may need GNU compilers (GCC 12+ recommended), a
```

   * - Gaea-C6 
     - /gpfs/f6/bil-fire8/world-shared/containers
   * - Orion/Hercules
     - /work/noaa/epic/role-epic/contrib/containers
   * - Derecho [#fn1]_
     - /glade/work/epicufsrt/contrib/containers
   * - NOAA Cloud [#fn2]_
     - /contrib/EPIC/containers

.. [#fn1] Software-stack container support on Derecho is still a work in progress.
.. [#fn2] NOAA Cloud containers are accessible only to users with EPIC resources through
   the Parallel Works dashboard.

Use one of the following approaches to define the ``IMG`` variable as the full
path to the container image that will be used for the SRW App build and runtime
workflow

.. code-block:: console

   # For GNU-based container image define:
   export IMG=<full-container-path>/rocky9-gcc13-ss192-ompi416.sif
   # for Intel-based container image define:
   export IMG=<full-container-path>/rocky9-oneapi2024.2-ss192.sif

Proceed to :ref:`downloading the SRW and submodules <DownloadSRWC>`.

**Option 2: Build a GNU-based container from Docker Hub**

If a staged GNU image is not available, build a Singularity/Apptainer image from
Docker Hub. 

.. code-block:: console

   singularity build rocky9-gcc13-ss192-ompi416.sif \

>       docker://noaaepic/rocky9-gcc13.3.1-spack-stack:v1.9.2-ufs-env-ompi416

Reviewer 2: A lot of your code blocks are spilt like this, is there a specific reasoning behind that?

Author: @EdwardSnyder-NOAA - no, no specific reason for splitting! Is it better to keep it as a single line and let the code block handle long lines?

Reviewer 2: The code block can handle long lines but I think it is fine as is.

Singularity/Apptainer image.

The examples in the following steps use local names for images and sandboxes.
In general, use the full path to each image or sandbox unless a specific step
instructs otherwise.

#. Create a writable sandbox from the Docker Hub image.
   Include bind-mounting host directories into the
   container. At a minimum, bind the top-level filesystem that contains your
   current directory, ``</top_dir>``, and any additional directories, ``/bind_add``,
   required for container builds. These may include system-dependent temporary
   build space, scratch space used as the default ``/tmp``, or ``/local``.
   Each bind path must be listed with a preceding ``-B`` flag. Typical bind
   directories for supported NOAA RDHPC Tier 1 platforms are listed in 
   :numref:`ContainerBindDirectoriesTable`.

   .. code-block:: console

      singularity build -B </top_dir> -B </bind_add> --sandbox --fix-perms rocky9-oneapi2024.2-ss192 \
         docker://noaaepic/rocky9-oneapi2024.2-spack-stack:v1.9.2-ufs-wm-env

#. Copy the helper scripts out of the sandbox.

   .. code-block:: console

      singularity exec rocky9-oneapi2024.2-ss192 cp /opt/intel-sandbox.sh .

>       singularity exec rocky9-oneapi2024.2-ss192 cp /opt/compilers_cp.sh .

Reviewer 2: Just curious what was the reasoning to have two separate scripts here, and not one, or even just a couple of commands listed in the document for the end user to run?

Author: @EdwardSnyder-NOAA -
This was intentional, to give users an idea what scripts are being staged locally. An alternative could be to name the scripts in the preceding text, and to make it a single-line command.

E.g.:

Copy the helper scripts, *intel-sandbox.sh* and *compilers_cp.sh* out of the sandbox.

```
      singularity exec rocky9-oneapi2024.2-ss192 cp /opt/*.sh .
      
```

   These scripts retrieve the Intel compiler and MPI components and reinstall
   them for use with the software-stack sandbox.

#. Create a sandbox with the original Intel oneAPI compilers by running the
   ``intel-sandbox.sh`` script from the same directory that contains the
   software-stack sandbox ``rocky9-oneapi2024.2-ss192``.

   .. code-block:: console

      ./intel-sandbox.sh

   After this step, an additional ``intel-sandbox`` sandbox container will be available.

#. Copy the required software and libraries from ``intel-sandbox`` to the
   original software-stack sandbox by running the ``compilers_cp.sh`` script.
   Include only the names of the source sandbox, ``intel-sandbox``, and target
   sandbox, ``rocky9-oneapi2024.2-ss192``; do not provide their full paths.

   .. code-block:: console

>       ./compilers_cp.sh intel-sandbox rocky9-oneapi2024.2-ss192

Reviewer 2: For some reason, on Orion, the intel directory in the intel-sandbox had no user permissions, so I had trouble copying the compilers over. Building the Intel sandbox on PW Azure worked fine though. Just curious if you tested building the Intel container on Orion and if you ran into any issues if you did?

Author: The script `intel-sandbox.sh` specifically uses `--fix-perms` option to build the sandbox with corrected permission for a user to read and access files and directories. If this does not help to view/read the directories on Orion, these could be due to restrictions of the Singularity container software with its present installation configuration - it is more restrictive than Apptainer as to building images/sandboxes. 
Just curious what the issue was with reading the `intel-sandbox` directories, anything specific? (For the sake of future troubleshooting of possible future issues)

For Orion and Hercules in particular, the workaround I used was to build container images or sandboxes using Apptainer, not Singularity. The trick is that the way updated Apptainer is installed on Orion and Hercules prevents it from integrating easily with other modules, so it can only be loaded in a separate environment (It could well be a deliberate installation configuration by security reasons).  

Loading the Apptainer on Orion/Hercules could be done as follows:
```
module load spack-managed-x86-64_v3/v1.0 apptainer/1.3.3
```
(in case of any troubles, do `module reset` prior to loading these modules)

After you built your image/sandbox, you could reset the module environment and load the Singularity module, which could be combined with other host modules as needed for the runtime. 

I could mention this in the documentation; however, the intent was that users of Tier 1 RDHPCs have the container images staged in a common location.

Suggested added note to address the case of Orion/Hercules:
```
.. note::

   Site-specific SingularityCE installations may restrict image and sandbox
   builds more than Apptainer installations. These restrictions are configured
   by system administrators for security reasons and therefore vary by system.
   If users encounter errors when building images or sandboxes with
   SingularityCE, they should use Apptainer for the build step when it is
   available. SingularityCE can still be used later to run the completed
   container image.

   On Orion/Hercules, for example, users who want to build the Intel sandbox
   and then create a combined sandbox that includes the full software stack and
   spack-stack libraries can load Apptainer with:

   .. code-block:: shell

      module load spack-managed-x86-64_v3/v1.0 apptainer/1.3.3

   This Apptainer module is loaded in a specific module environment that is
   not easily combined with other host modules. However, this environment is
   needed only while building the sandbox or image. After the final combined
   image has been created, the Apptainer module environment can be unloaded, and
   the more readily available SingularityCE module can be used for container
   runtime.
```

   host directories as required.

   .. code-block:: console

      singularity build -B /<top-level-dir> --fix-perms rocky9-oneapi2024.2-ss192.sif \
         rocky9-oneapi2024.2-ss192

   After the image is built successfully, the sandbox can be removed. Finally, 
   define the ``IMG`` variable for use in later steps.
   
   .. code-block:: console

      export IMG=<full-container-path>/rocky9-oneapi2024.2-ss192.sif
   
Proceed with downloading, building, and running the SRW App.

.. _DownloadSRWC:

Download the UFS SRW App and Submodules
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Clone the UFS SRW App develop branch from the GitHub repository as is done
when  :ref:`Building the SRW App <BuildSRW>`.

.. include:: ../../doc-snippets/clone.rst

.. code-block:: console

   cd ufs-srweather-app

.. include:: ../../doc-snippets/externals.rst

Save the environment variable SRW for later use:

.. code-block:: console

   export SRW=${PWD}

.. _ShellInteractiveC:

Shell into the Software-Stack Container
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Open an interactive shell inside the container before building the SRW App. Bind

> the top-level host filesystem that contains the SRW App checkout, the input

Reviewer 2: I don't think checkout is needed here.

Author: A better wording suggested, with additional explanations:

"Bind-mount all host filesystems and directories that contain the SRW App, input data, and experiment directories. These filesystems are identified during the build stage and recorded in the runtime environment files so that the required paths are accessible inside the container at runtime."

Author: @EdwardSnyder-NOAA - let me know if you think this works better!

Reviewer 2: Updates look good to me.

Author: Will add them to the next commit!

.. list-table:: Locations of pre-built container images on supported systems
   :widths: 20 50
   :header-rows: 1

   * - Machine
     - File Location
   * - Ursa
     - /scratch3/NCEPDEV/nems/role.epic/containers
   * - Gaea-C6 
     - /gpfs/f6/bil-fire8/world-shared/containers
   * - Orion/Hercules
     - /work/noaa/epic/role-epic/contrib/containers
   * - Derecho [#fn1]_
     - /glade/work/epicufsrt/contrib/containers
   * - NOAA Cloud [#fn2]_
     - /contrib/EPIC/containers

.. [#fn1] Software-stack container support on Derecho is still a work in progress.
.. [#fn2] NOAA Cloud containers are accessible only to users with EPIC resources through
   the Parallel Works dashboard.

Use one of the following approaches to define the ``IMG`` variable as the full
path to the container image that will be used for the SRW App build and runtime
workflow

.. code-block:: console

   # For GNU-based container image define:
   export IMG=<full-container-path>/rocky9-gcc13-ss192-ompi416.sif
   # for Intel-based container image define:
   export IMG=<full-container-path>/rocky9-oneapi2024.2-ss192.sif

> Proceed to :ref:`downloading the SRW an submodules <DownloadSRWC>`.

Reviewer: ```suggestion
Proceed to :ref:`downloading the SRW and submodules <DownloadSRWC>`.
```

**Option 2: Build a GNU-based container from Docker Hub**

If a staged GNU image is not available, build a Singularity/Apptainer image from
Docker Hub. 

.. code-block:: console

   singularity build rocky9-gcc13-ss192-ompi416.sif \
      docker://noaaepic/rocky9-gcc13.3.1-spack-stack:v1.9.2-ufs-env-ompi416

   export IMG=${PWD}/rocky9-gcc13-ss192-ompi416.sif

If the build fails because of cache or temporary-space limits, you may need to clear
the default temporary directories (e.g, *${HOME}/.singularity/cache* or
*${HOME}/.apptainer/cache*) and proceed to allocate more temporary space as 
outlined in :ref:`Appendix`.

> After the image is built, proceed to :ref:`downloading the SRW an submodules <DownloadSRWC>`.

Reviewer: ```suggestion
After the image is built, proceed to :ref:`downloading the SRW and submodules <DownloadSRWC>`.
```

the following line to ``wflow_hercules.lua`` or ``wflow_orion.lua``:

.. code-block:: lua

   load("singularity")

.. note::

   The exact container runtime module may vary by platform. Use the module name
   provided by the target system, such as ``singularity`` or ``apptainer``.

.. _PrepareConfigurationFilesC:

Prepare Configuration Files
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Adapt the configuration files for the target platform and for the community
test case.

First, prepare the main SRW App configuration file:

.. code-block:: console

   cd ../ush
   cp config.container.yaml config.yaml

Edit ``./ush/config.yaml`` and set the following variables as needed:

.. code-block:: yaml

   ACCOUNT: epic
   COMPILER: gnu
   USE_CRON_TO_RELAUNCH: false
   EXPT_SUBDIR: test_community

Set ``ACCOUNT`` to the account or project name used on the target system. Set
``COMPILER`` to the compiler used by the container software stack, such as
``gnu`` or ``intel``. Set ``USE_CRON_TO_RELAUNCH`` to ``true`` only on systems
where cron-based relaunching is allowed. Modify ``EXPT_SUBDIR`` if a different
experiment directory name is desired.

Next, prepare the machine file:

.. code-block:: console

>    cd machines

Reviewer: ```suggestion
   cd machine
```

   cp <platform>.yaml container.yaml

For example:

.. code-block:: console

   cp ursa.yaml container.yaml

> Edit ``./ush/machines/container.yaml`` for the container workflow. Modify

Reviewer: ```suggestion
Edit ``./ush/machine/container.yaml`` for the container workflow. Modify
```
